torch_geometric/loader/link_neighbor_loader.py

```diff
@@ -34,37 +32,17 @@ def __init__(
         # TODO if self.edge_time is not None and
         # `src` or `dst` nodes don't have time attribute
         # i.e node_time_dict[input_type[0/-1]] doesn't exist
-        # set it to largest representabel torch.long.
-        self.num_src_nodes = num_src_nodes
-        self.num_dst_nodes = num_dst_nodes
-
-        if self.num_src_nodes is None or self.num_dst_nodes is None:
-            if self.data_cls == 'custom':
-                _, graph_store = data
-                edge_attrs = graph_store.get_all_edge_attrs()
-                edge_types = [attr.edge_type for attr in edge_attrs]
-
-                # Edge label index is part of the graph:
-                if self.input_type in edge_types:
-                    self.num_src_nodes = graph_store.num_src_nodes(
-                        edge_type=self.input_type)
-                    self.num_dst_nodes = graph_store.num_dst_nodes(
-                        edge_type=self.input_type)
-                else:
-                    # We do not support querying the number of nodes by the
-                    # feature store, so we throw an error here:
-                    raise ValueError(
-                        f"Use of a remote backend with "
-                        f"{self.__class__.__name__} requires the "
-                        f"specification of source and destination nodes, as "
-                        f"the edge label index {self.input_type} is not part "
-                        f"of the specified graph.")
-
-            elif issubclass(self.data_cls, Data):
-                self.num_src_nodes = self.num_dst_nodes = data.num_nodes
-            else:  # issubclass(self.data_cls, HeteroData):
-                self.num_src_nodes = data[self.input_type[0]].num_nodes
-                self.num_dst_nodes = data[self.input_type[-1]].num_nodes
+        # set it to largest representable torch.long.
+        if self.data_cls == 'custom':
+            feature_store, graph_store = data
+            self.num_src_nodes, self.num_dst_nodes = remote_backend.num_nodes(
+                feature_store, graph_store, self.input_type)
+
+        elif issubclass(self.data_cls, Data):
+            self.num_src_nodes = self.num_dst_nodes = data.num_nodes
+        else:  # issubclass(self.data_cls, HeteroData):
+            self.num_src_nodes = data[self.input_type[0]].num_nodes
+            self.num_dst_nodes = data[self.input_type[-1]].num_nodes
```

Comment:
This can also use the `remote_backend`, right?
Ok if its refactored in another PR.

Reply:
In theory, yes - but I think it's best to start consolidating Data/HeteroData logic behind FeatureStore/GraphStore logic in a separate PR.